src/lib/hooks/routing/clientSideSmartOrderRouter.ts

```diff
@@ -19,50 +21,45 @@ export function isSupportedChainId(chainId: ChainId | undefined): boolean {
 
 async function getQuote(
   {
-    type,
+    tradeType,
     tokenIn,
     tokenOut,
     amount: amountRaw,
   }: {
-    type: 'exactIn' | 'exactOut'
+    tradeType: TradeType
     tokenIn: { address: string; chainId: number; decimals: number; symbol?: string }
     tokenOut: { address: string; chainId: number; decimals: number; symbol?: string }
     amount: BigintIsh
   },
   router: AlphaRouter,
-  config: Partial<AlphaRouterConfig>
-): Promise<{ data: GetQuoteResult; error?: unknown }> {
-  const currencyIn = new Token(tokenIn.chainId, tokenIn.address, tokenIn.decimals, tokenIn.symbol)
-  const currencyOut = new Token(tokenOut.chainId, tokenOut.address, tokenOut.decimals, tokenOut.symbol)
+  routerConfig: Partial<AlphaRouterConfig>
+): Promise<QuoteResult> {
+  const tokenInIsNative = Object.values(SwapRouterNativeAssets).includes(tokenIn.address as SwapRouterNativeAssets)
+  const tokenOutIsNative = Object.values(SwapRouterNativeAssets).includes(tokenOut.address as SwapRouterNativeAssets)
 
-  const baseCurrency = type === 'exactIn' ? currencyIn : currencyOut
-  const quoteCurrency = type === 'exactIn' ? currencyOut : currencyIn
-  const amount = CurrencyAmount.fromRawAmount(baseCurrency, JSBI.BigInt(amountRaw))
+  const currencyIn = tokenInIsNative
+    ? nativeOnChain(tokenIn.chainId)
+    : new Token(tokenIn.chainId, tokenIn.address, tokenIn.decimals, tokenIn.symbol)
+  const currencyOut = tokenOutIsNative
+    ? nativeOnChain(tokenOut.chainId)
+    : new Token(tokenOut.chainId, tokenOut.address, tokenOut.decimals, tokenOut.symbol)
 
-  const swapRoute = await router.route(
-    amount,
-    quoteCurrency,
-    type === 'exactIn' ? TradeType.EXACT_INPUT : TradeType.EXACT_OUTPUT,
-    /*swapConfig=*/ undefined,
-    config
-  )
+  const baseCurrency = tradeType === TradeType.EXACT_INPUT ? currencyIn : currencyOut
+  const quoteCurrency = tradeType === TradeType.EXACT_INPUT ? currencyOut : currencyIn
 
-  if (!swapRoute) throw new Error('Failed to generate client side quote')
+  const amount = CurrencyAmount.fromRawAmount(baseCurrency, JSBI.BigInt(amountRaw))
+  const swapRoute = await router.route(amount, quoteCurrency, tradeType, /*swapConfig=*/ undefined, routerConfig)
 
-  return { data: transformSwapRouteToGetQuoteResult(type, amount, swapRoute) }
-}
+  // this check is intentionally left delayed until after the `router.route` call so that on empty input, the
+  // client side routing will begin initialization
```

Comment:
I think I need a better explanation here - what does it mean that "client side routing will begin initialization"? AFAIK we're not doing any work here, we're just returning state. What kind of "initialisation" do you mean?

Reply:
@zzmp can you help me with wording here?

Reply:
What if you simplify this by making it explicit in code.

```
const isInitialization = !amountRaw
if (isInitialization) {
  // Initialization should use a dummy amount so that the router can initialize.
  // The router makes on-chain calls to initialize a route. Calling `router.route` to initialize
  // significantly improves the lag for the first *actual* route.
  const amount = CurrencyAmount.fromRawAmount(baseCurrency, JSBI.BigInt(1))
  router.route(amount, quoteCurrency, tradeType, /*swapConfig=*/ undefined, routerConfig)
  return { state: QuoteState.INITIALIZED }
}
```